### ISSUE TITLE
Question: How to install certificate

### ISSUE THREAD

**Issue by anti-matter** [justcoding121/Titanium-Web-Proxy#478]
Problem: I want to deploy the proxy server on a remote server and have clients configure their proxy settings to go through that proxy. Ideally, The clients should be able to just enter the proxy info in their proxy config and hot have to manually install a certificate on the client. Are there concrete examples of how to do this?

I've seen another issue here that describes the same problem, but in that case the user didn't care about inspecting the traffic, so the solution was to just not decrypt the ssl. In my case, that won't work as I need to inspect the traffic and take specific action(s).

Thanks!

**Comment by anti-matter** [comment]
Update: I decided to just forego the certificate issue at this point and attempt to get the proxy server running on a Windows Server. The problem I've been having so far with getting the proxy running is that all the proxyserver.CertificateManager.XXXCertificate() functions fail with the error that the system cannot access the registry.

So, I tried not supporting SSL at all as per another issue posted here, but even though I commented out all the certificate trust/creation lines, I still get the same error which leads me to believe the proxy is still trying to trust/create a root certificate.

Here's my init code:
```C#
            try
            {
                //locally trust root certificate used by this proxy
                //Proxy.CertificateManager.TrustRootCertificate();

                //optionally set the Certificate Engine
                //Under Mono only BouncyCastle will be supported
                //Proxy.CertificateManager.CertificateEngine = Network.CertificateEngine.BouncyCastle;

                Proxy.BeforeRequest += OnRequest;
                Proxy.BeforeResponse += OnResponse;
                //Proxy.ServerCertificateValidationCallback += OnCertificateValidation;
                //Proxy.ClientCertificateSelectionCallback += OnCertificateSelection;

                Proxy.ProxyAuthenticationSchemes = new[] { "basic" };
                Proxy.ProxyBasicAuthenticateFunc = ProxyBasicAuthenticateFunc;

                _explicitProxyEndPoint = new ExplicitProxyEndPoint(IPAddress.Any, 8000, false);
                // ReSharper disable once RedundantEmptyObjectOrCollectionInitializer
                {
                    //Use self-issued generic certificate on all https requests
                    //Optimizes performance by not creating a certificate for each https-enabled domain
                    //Useful when certificate trust is not required by proxy clients
                    //GenericCertificate = new X509Certificate2(Path.Combine(System.IO.Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location), "genericcert.pfx"), "password")
                };

                //Fired when a CONNECT request is received
                _explicitProxyEndPoint.BeforeTunnelConnectRequest += OnBeforeTunnelConnect;

                //An explicit endpoint is where the client knows about the existence of a proxy
                //So client sends request in a proxy friendly manner
                Proxy.AddEndPoint(_explicitProxyEndPoint);
                Proxy.Start();

                //Proxy.UpStreamHttpProxy = new ExternalProxy() { HostName = "localhost", Port = 8888 };
                //Proxy.UpStreamHttpsProxy = new ExternalProxy() { HostName = "localhost", Port = 8888 };

                foreach (var endPoint in Proxy.ProxyEndPoints)
                {
                    _logger.LogInformation(
                        $"Listening on '{endPoint.GetType()}' endpoint at IP {endPoint.IpAddress} and port: {endPoint.Port}");
                }
            }
            catch (Exception ex)
            {
                LogAsync(LogLevel.Warning, "Failed to initialize root cert: {0}", ex.Message).Wait();
                return;
            }
```

Is Titanium-Web-Proxy supported/tested on server at all?

Thanks!

**Comment by justcoding121** [comment]
I haven't tested on windows server. If you have a root certificate you can use it and trust manually by installing the certificate in server. You can try bouncy castle if windows certificate generation seem to be the problem. 

`Proxy.CertificateManager.CertificateEngine = Network.CertificateEngine.BouncyCastle;`

To use your own root certificate see below property. Load the certificate and set the property before straying the proxy. Clients would need to also install the root (without private key is okay for clients) or an intermediary  certificate signed by root.

https://justcoding121.github.io/Titanium-Web-Proxy/docs/api/Titanium.Web.Proxy.Network.CertificateManager.html#Titanium_Web_Proxy_Network_CertificateManager_RootCertificate

**Comment by anti-matter** [comment]
When running on Windows Server 2016, here's the exception I'm getting:

[8/11/2018 6:36:08 PM] Warning: at Microsoft.Win32.ThrowHelper.ThrowSecurityException(String msg)
at Microsoft.Win32.RegistryKey.InternalOpenSubKeyCore(String name, Boolean writable, Boolean throwOnPermissionFailure)
at Microsoft.Win32.RegistryKey.OpenSubKey(String name, Boolean writable)
at Titanium.Web.Proxy.Helpers.SystemProxyManager.GetProxyInfoFromRegistry()
at Titanium.Web.Proxy.ProxyServer.Start()

The server process I'm running (a asp.net core app running in IIS) does not have access to the registry.

Here is the code from ProxyServer.cs which is the reason the proxy is attempting to access the registry:

```C#
// clear any system proxy settings which is pointing to our own endpoint (causing a cycle)
            // due to ungracious proxy shutdown before or something else
            if (systemProxySettingsManager != null && RunTime.IsWindows)
            {
                var proxyInfo = systemProxySettingsManager.GetProxyInfoFromRegistry();
                if (proxyInfo.Proxies != null)
                {
                    var protocolToRemove = ProxyProtocolType.None;
                    foreach (var proxy in proxyInfo.Proxies.Values)
                    {
                        if (NetworkHelper.IsLocalIpAddress(proxy.HostName)
                            && ProxyEndPoints.Any(x => x.Port == proxy.Port))
                        {
                            protocolToRemove |= proxy.ProtocolType;
                        }
                    }

                    if (protocolToRemove != ProxyProtocolType.None)
                    {
                        systemProxySettingsManager.RemoveProxy(protocolToRemove, false);
                    }
                }
            }
```

**Comment by justcoding121** [comment]
Is is unimportant, and you can wrap the above lines with try catch ignore if you need an immediate fix with your own build of the project. Or wait until next release when this is fixed.

**Comment by anti-matter** [comment]
What I did was add a flag (a property) in my own build which is used to determine whether to support proxy client changes, or not. The reason I use a property flag is because during dev builds I want the functionality of enabling the client for testing. When running in production, I set the flag to disable all proxy client config settings.

BTW, there ended up NOT begin any issue with the certificates on Server. The current mechanism works just fine. The issue was with me. :)

Thank you!
